### PR TITLE
Fix make disk image automatic size

### PR DESCRIPTION
**Pull request by samueldr**
###### Motivation for this change

 - #89331
 - #107212
 - #120481

This is doing the actual accounting we need to do to make the partition big enough to fit data.

This was tested by setting `additionalSpace` to `0`. Any other values (like the default 512MiB) can hide subtle issues. But the default 512MiB is fine, since it also hardens this implementation from small deviations.

The accounting fixes here comes from work I did for Mobile NixOS.

 - https://github.com/NixOS/mobile-nixos/tree/master/lib/image-builder

See also:

 - https://github.com/NixOS/nixpkgs/blob/93de027ee045d0c1fbf13418fb86d86e6309b962/nixos/lib/make-ext4-fs.nix#L58-L63

Don't forget to read the commit messages!

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mic92** [comment]
I guess this could also revert the revert if it works? https://github.com/NixOS/nixpkgs/pull/120487
I could test this on zfs with compression enabled.

**Review by lukegb — APPROVED** [review]

**Comment by lukegb** [comment]
Yeah, this should probably revert the revert too - or, probably better, re-cherry-pick f3aa040bcbf39935e7e9ac7a7296eac9da7623ec?

**Review by Lassulus — APPROVED** [review]
code looks very clean and is good documented, thanks a lot for fixing my mess :)

I would be very happy to have the dynamic sized images in 21.05 if everything works

**Comment by samueldr** [comment]
Since it could get confusing "how did you test this if the commits were already reverted?"

Well, I was based off a commit from before the reverts merge PR :).

**Comment by Mic92** [comment]
I tried to build openstack with nixos-generators, but it segfaulted my nix. This is probably unrelated to changes here.

**Comment by numinit** [comment]
@Mic92 I ran into the same thing. I just happened to be concurrently refactoring make-disk-image.nix to allow for creating swap space after the boot partition and rebased against this PR. Tried `nix-shell -p nix` to get the non-experimental nix and everything worked afterwards.

Anyway, this change looks good to me after spending a Saturday afternoon familiarizing myself with this part of nixpkgs. :+1:

[The swap space changes are in preparation for trying to add ZFS support to it too - my idea is automatically provisioning NixOS images that look [like this](https://grahamc.com/blog/nixos-on-zfs). I have lots of servers where the dataset layout @grahamc  describes works _really_ well, and would love to just build an image that I can dd straight to the disk instead of running a long playbook of commands over and over again for every server with minor differences for every one...]

**Comment by samueldr** [comment]
@numinit WAIT! collaborate on #106574 !

**Comment by numinit** [comment]
@samueldr Nice. Didn't know about that one. I've not done any work on the ZFS stuff yet, but will still PR the swapspace changes because they're useful anyway here. :-)

File: nixos/lib/make-disk-image.nix
Comment (numinit) [inline nixos/lib/make-disk-image.nix:288]: This isn't necessarily a multiple of 512 bytes and makes fdisk/gdisk complain. parted seems to ignore it, but probably still a good idea to round up to the next highest blockSize here.

Comment (samueldr) [reply nixos/lib/make-disk-image.nix:288]: This probably would be wrong.

What I assume it complains about is the partitions starting/ending unaligned, and not necessarily on sector size (commonly 512 bytes). I believe the alignment by default is 1MiB? Though alignment is actually a property of the block device being partitioned! Doing any alignment on an image might actually end up wrong!

For this, we'd need to re-do the whole disk image build logic to actually work with discrete filesystem images that we *then* place as partitions into a larger disk image.

Right now, we're actually trying to carve a hole into a bigger image we *hope* we're making big enough. Well, we're doing less of it with this PR.

Though I guess rounding up to the nearest megabyte is a clean way to avoid this warning.

Comment (numinit) [reply nixos/lib/make-disk-image.nix:288]: It's actually complaining about file size (three whole times for some reason). And it's just gdisk.

```
$ truncate -s 1234 test
$ gdisk test
GPT fdisk (gdisk) version 1.0.5

Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
```

Yeah, rounding up to the nearest MB would fix it. I think parted actually recommends that everything be 4MiB aligned at least. Best we can do is guess here anyway and 1 MB alignment seems good enough for most cases.

Comment (samueldr) [reply nixos/lib/make-disk-image.nix:288]: Right, this warning is probably a non-issue if it is on a disk image:

 - https://sourceforge.net/p/gptfdisk/code/ci/eae9f7e2e1470ed2d5bc029180aa2adc88fbc4a7/tree/diskio-unix.cc#l496
 - https://sourceforge.net/p/gptfdisk/code/ci/eae9f7e2e1470ed2d5bc029180aa2adc88fbc4a7/tree/diskio-unix.cc#l493

Hmm... Though it would mean that the backup GPT at the end of the disk is not at LBAEND-33, but rather LBAEND-34, since it rounded down the amount of sectors. BUT, again this is a non-issue since when `dd`'ing the disk image to the actual device, the device will always be bigger, and that the backup GPT will invariably not be at the end sector.

Comment (samueldr) [reply nixos/lib/make-disk-image.nix:290]: And it is done

https://github.com/NixOS/nixpkgs/pull/120489/commits/7b8b3fab6dce5357c14fc435b436514fc985f0f7

File: nixos/lib/make-disk-image.nix
Comment (numinit) [inline nixos/lib/make-disk-image.nix:281]: Do we need to add size for directories (plus their entries) and symlinks, or do you think this is a "good enough" estimate, those aside?

Comment (samueldr) [reply nixos/lib/make-disk-image.nix:281]: This inodes accounting comes from a contributor who's not contributing anymore. I don't know the details. But I assume they take as much space as files. Directories and symlinks have distinct inodes when listing them with `ls -i`.

**Comment by vcunat** [comment]
Some of you might know... [one image job](https://hydra.nixos.org/build/142244712) now started blocking `nixos-unstable` channel due to producing a too large output (slightly over the 2 GiB limit).

**Comment by samueldr** [comment]
Last successful:

```
irb(main):008:0> 3128950784.0 / 1024 / 1024 / 1024
=> 2.9140625
irb(main):009:0> 3128950784.0 / 1000 / 1000 / 1000
=> 3.128950784
```

```
irb(main):002:0> 3128950784.0 / 1024 / 1024 / 1024
=> 2.9140625
irb(main):003:0> 3128950784.0 / 1000 / 1000 / 1000
=> 3.128950784
```

So uh... Odd!

The **reported** disk image size is the same from the changes from this PR.

It might be that the actual file being produced is different.

**Comment by samueldr** [comment]
```
[20:25:29] <samueldr> max output size is 2147483648
[20:27:16] <samueldr> is that once the NAR is compressed?
[20:27:59] <lukegb> yes
[20:31:26] <samueldr> right
[20:31:48] <samueldr> still, odd that it's over only sometimes?
[20:32:34] <lukegb> I guess building the image is non-deterministic
[20:32:41] <lukegb> and we're _just_ close to the limit
[20:34:58] <samueldr> ~1705414492.16 bytes
[20:35:28] <samueldr> I would be surprised that it is *that much* irreproducible
[20:35:42] <samueldr> output size from https://hydra.nixos.org/build/142182234#tabs-details
```

**Comment by Lassulus** [comment]
hmm, I guess we can set the filesize of the amazon image to unblock the channel? maybe we can set it to auto again when we reduce the size of the image somehow of raise the limit?

**Comment by vcunat** [comment]
Unblocked by PR #121352.

**Pull request merged**
